File: pipeline/src/workflows/project_collaboration.py
"""
================================================================================
Workflow: ProjectCollaboration  (the generic project/opportunity HITL reaction)
================================================================================

THE CANONICAL PATTERN. This is the ONE generic, overlay-parameterized reaction
template that new project/opportunity automation should reuse — not another
bespoke On<Event> class. It generalizes OnProposalAdvancedToReview (park a human
gate, nudge it, resume on completion) and is launched the OnCmsContentRequested
way (emit its trigger event with an OVERLAY payload that carries the specifics).

TRIGGER:  proposal:project.collaboration_requested:single
          phase='single' ⇒ launchable on demand via lib/process/launch-template.ts
          (and by any bridge that emits the event — the create route for V1, the
          Stripe webhook when billing is live). Condition: the overlay names a
          proposal OR an opportunity to act on.

WHAT IT DOES:  parks a single, fully payload-parameterized human gate (a `tasks`
          ledger row: assignee, title, entity, nudge cadence, due) and resumes
          when that task is completed (or force-advanced, or a wait_for event
          fires). On resume it announces completion. It is a TRANSIENT reaction:
          park → resume → complete; never a weeks-long `running` row.

READS, NEVER WRITES, STATUS.  The overlay carries `stage` (and `proposalId`) for
          context and the task entity; this template references it but NEVER
          writes proposals.stage — advanceProposalStage stays the sole authority.
          The instance reacts to status; it does not own or drive it.

OVERLAY (payload) — frozen at launch, immutable forward (force-advance aside):
    proposalId        (uuid, optional)  — the project, when project-scoped
    opportunityId     (uuid, optional)  — the spine key; written to
                                          process_instances.opportunity_id so the
                                          control tower can chain/roll up by opp
    scope             ('opp'|'spotlight'|'project'|'contract') — written to
                                          process_instances.scope (the spine
                                          discriminator); defaults null if absent
    tenantId          (uuid, optional)  — the owning tenant (admin gate ⇒ null)
    stage             (text, optional)  — the project stage this gate sits at (context)
    taskType          (text, REQUIRED)  — the gate kind, e.g. "admin_review"
    taskTitle         (text)            — human-readable gate title
    assigneeRole      (text)            — role bucket, e.g. "rfp_admin"
    assigneeUser      (uuid, optional)  — a specific assignee (overrides the role)
    entityType        (text)            — e.g. "proposal" | "opportunity"
    entityRef         (uuid)            — the entity the assignee acts on
    nudgeDays         (int[])           — relative nudge cadence, e.g. [1,3,5]
    dueMinutes        (int)             — task due window (drives urgency + nudges)
    parkMinutes       (int, optional)   — instance park ceiling (the abandon
                                          backstop); overlay-driven so a weeks-long
                                          customer-build gate is not capped at the
                                          static default. Falls back to the step
                                          timeout (a generous 30-day backstop).
    completeTemplate  (text, optional)  — email template on completion; absent ⇒
                                          no email (system.notify degrades cleanly)

DESIGN NOTES:
    - One gate per instance, by design. The lifecycle has up to ~3 human gates
      (V1 → Locked), but each is driven by its OWN event (a stage advance, a
      purchase, a pin) — so the faithful model is one transient reaction per
      gate, launched reactively, NOT one long instance pre-chaining all stages
      (which would fight the gate machine's authority). Multi-gate = multi-launch.
    - assignee/title/entity/nudges/due are all resolved per-instance from the
      overlay via the same resolve_input paths as input_map (_create_task) — the
      static step stays generic; the payload carries the specifics.
    - The gate resumes via task completion (no wait_for needed). A reviewer who
      rejects force-fails the instance from the Process Ledger instead.
    - parkMinutes: the TASK's due_at/nudges (dueMinutes/nudgeDays) drive the human
      cadence; parkMinutes is only the instance's abandon backstop. They are
      decoupled on purpose — see manager.execute_instance.
================================================================================
"""
from workflows.base import Workflow, Step, StepType, EventTrigger

# Instance park ceiling (abandon backstop) when the overlay omits parkMinutes:
# 30 days, generous enough that a normally-progressing gate never trips the
# paused-deadline sweep, while a truly-abandoned gate is eventually reaped.
_PARK_BACKSTOP_MINUTES = 43200


class ProjectCollaboration(Workflow):
    description = (
        "Generic project/opportunity HITL reaction: park one payload-parameterized "
        "human gate (tasks row + nudges), resume on completion, then notify. Launched "
        "with an overlay; reads proposals.stage, never writes it."
    )

    # NOTE: scope is NOT a class attribute. It is inherently per-launch (a purchase
    # bridge launches scope='opp'; the create-route launches scope='project'), so
    # create_instance reads it from the overlay (validated against the mig-088 CHECK).
    # The launcher MUST set payload.scope; absent ⇒ the instance is unscoped (null).

    trigger = EventTrigger(
        namespace="proposal",
        type="project.collaboration_requested",
        phase="single",
        # A launch must name something to act on — guards junk launches.
        condition=lambda p: bool(p.get("proposalId") or p.get("opportunityId")),
    )

    steps = [
        Step(
            name="collaborate",
            step_type=StepType.TODO,
            action="todo",
            # Every field resolves from the overlay (payload.X) per-instance.
            task_type="payload.taskType",
            task_title="payload.taskTitle",
            assignee_role="payload.assigneeRole",
            assignee_user="payload.assigneeUser",
            entity_type="payload.entityType",
            entity_ref="payload.entityRef",
            nudge_days="payload.nudgeDays",
            due_in_minutes="payload.dueMinutes",
            # Instance park ceiling. The overlay can extend it via payload.parkMinutes
            # (handled in manager.execute_instance); this static value is the fallback
            # abandon backstop, NOT the human-facing due date (that's dueMinutes).
            timeout_minutes=_PARK_BACKSTOP_MINUTES,
        ),
        Step(
            name="notify_done",
            step_type=StepType.NOTIFY,
            action="system.notify",
            depends_on="collaborate",
            # completeTemplate absent ⇒ system.notify degrades to no-email (never a
            # hard failure), so a gate that needs no completion email just omits it.
            input_map={
                "channel": '"email"',
                "template": "payload.completeTemplate",
                "proposal_id": "payload.proposalId",
                "tenant_id": "payload.tenantId",
            },
        ),
    ]

File: frontend/lib/process/project-collaboration.ts
/**
 * launchProjectCollaboration — the CANONICAL way to start a project/opportunity
 * HITL reaction from the frontend. New project automation calls THIS, not a
 * bespoke flow: it builds a complete, guarded overlay for the generic
 * `ProjectCollaboration` pipeline template (pipeline/src/workflows/
 * project_collaboration.py) and launches it via the existing launchTemplate
 * engine path (emit the single-phase trigger event with the overlay frozen as
 * the instance payload; the pipeline parks a `tasks` gate and nudges it).
 *
 * Why a helper and not a raw launchTemplate call at each site: the generic
 * template resolves every gate field from the overlay, and a MISSING required
 * field (taskType/assigneeRole) silently degrades to a corrupt task. This helper
 * makes those fields REQUIRED in the type system and guards them at runtime, so
 * a bridge can never launch an unassigned/mis-typed gate. It also stamps the
 * spine keys (opportunityId + scope) the control tower rolls up by.
 *
 * Bridges that call this today: the proposal create route (scope='project', the
 * 72h admin-review gate) and the Stripe webhook on an opportunity purchase
 * (scope='opp', the workspace-setup gate) — closing the purchase.completed orphan.
 */
import { launchTemplate, type LaunchActor, type LaunchResult } from '@/lib/process/launch-template';
import { isValidUUID } from '@/lib/validation';

export type SpineScope = 'opp' | 'spotlight' | 'project' | 'contract';

export interface ProjectCollaborationLaunch {
  /** Who launched it. For a system bridge (webhook) set actorType:'system'. */
  actor: LaunchActor;
  actorType?: 'user' | 'system';
  /** Owning tenant (null for an admin/platform gate). */
  tenantId: string | null;
  /** Spine discriminator — written to process_instances.scope. */
  scope: SpineScope;
  /** The immutable spine key — written to process_instances.opportunity_id. */
  opportunityId: string;
  /** The project, when project-scoped (the gate's stage context). */
  proposalId?: string | null;
  /** The project stage this gate sits at (context only; never written back). */
  stage?: string | null;

  // ── the gate (all REQUIRED-by-type so a launch can't write a corrupt task) ──
  taskType: string;
  taskTitle: string;
  assigneeRole: string;
  assigneeUser?: string | null;
  entityType: string;
  entityRef: string;
  nudgeDays?: number[];
  /** Task due window (drives urgency + nudges). Default 72h. */
  dueMinutes?: number;
  /** Instance park ceiling (abandon backstop). Omit ⇒ the template's 30-day default. */
  parkMinutes?: number;
  /** Email template on completion; omit ⇒ no email (degrades cleanly). */
  completeTemplate?: string | null;
}

const WORKFLOW = 'ProjectCollaboration';

export async function launchProjectCollaboration(
  opts: ProjectCollaborationLaunch,
): Promise<LaunchResult> {
  // Runtime guard: the generic template can't enforce required overlay fields
  // (a missing one degrades to a corrupt task: mis-typed, unassigned, or pointing
  // at nothing), so refuse an incomplete launch — matching this helper's promise
  // that a bridge "can never launch a corrupt gate". Covers every consumed field.
  const missing: string[] = [];
  if (!opts.taskType?.trim()) missing.push('taskType');
  if (!opts.taskTitle?.trim()) missing.push('taskTitle');
  if (!opts.assigneeRole?.trim()) missing.push('assigneeRole');
  if (!opts.entityType?.trim()) missing.push('entityType');
  if (!opts.entityRef?.trim()) missing.push('entityRef');
  if (!opts.opportunityId?.trim()) missing.push('opportunityId');
  if (missing.length > 0) {
    return {
      ok: false,
      status: 400,
      error: `ProjectCollaboration launch missing required field(s): ${missing.join(', ')}`,
      code: 'INCOMPLETE_OVERLAY',
    };
  }

  // UUID-format guard: entityRef becomes the task's entity_id and opportunityId
  // the spine roll-up key — both go through the pipeline's _safe_uuid, which nulls
  // a non-UUID. A presence-only check would let an operator-influenceable value
  // (e.g. raw Stripe metadata) pass, then SILENTLY null the entity + scope key —
  // a task pointing at nothing and an unscoped instance (re-creating the orphan
  // this bridge exists to close). Refuse loudly instead.
  const malformed: string[] = [];
  if (!isValidUUID(opts.entityRef)) malformed.push('entityRef');
  if (!isValidUUID(opts.opportunityId)) malformed.push('opportunityId');
  if (opts.assigneeUser && !isValidUUID(opts.assigneeUser)) malformed.push('assigneeUser');
  if (malformed.length > 0) {
    return {
      ok: false,
      status: 400,
      error: `ProjectCollaboration launch has non-UUID field(s): ${malformed.join(', ')}`,
      code: 'INVALID_OVERLAY',
    };
  }

  // Build the overlay — omit undefined keys so the template's resolve_input sees
  // a clean payload (and parkMinutes/completeTemplate fall back to their defaults).
  const overlay: Record<string, unknown> = {
    scope: opts.scope,
    opportunityId: opts.opportunityId,
    taskType: opts.taskType,
    taskTitle: opts.taskTitle,
    assigneeRole: opts.assigneeRole,
    entityType: opts.entityType,
    entityRef: opts.entityRef,
    // Days-BEFORE-due: [2,1] (48h + 24h out). A value ≥ the full due window (e.g. 3 on a 72h
    // gate) makes the first nudge threshold equal the creation instant — an immediate nudge.
    nudgeDays: opts.nudgeDays ?? [2, 1],
    dueMinutes: opts.dueMinutes ?? 4320, // 72h
  };
  if (opts.proposalId) overlay.proposalId = opts.proposalId;
  if (opts.stage) overlay.stage = opts.stage;
  if (opts.assigneeUser) overlay.assigneeUser = opts.assigneeUser;
  if (opts.parkMinutes != null) overlay.parkMinutes = opts.parkMinutes;
  if (opts.completeTemplate) overlay.completeTemplate = opts.completeTemplate;

  return launchTemplate({
    workflowName: WORKFLOW,
    overlay,
    actor: opts.actor,
    actorType: opts.actorType ?? 'user',
    tenantId: opts.tenantId,
  });
}
